### PR TITLE
SPOI-11028 update the systemalerts and security documents for SMTP configuration and keystore related content

### PR DESCRIPTION
**Pull request by sanjaypujare**
@amberarrow  pls review and merge. Also @wvary to review and comment

### CODE REVIEW

File: docs/dtgateway_security.md [hunk 1]
```diff
@@ -821,34 +821,70 @@ This command returns the information about the user “john”.
 
 This command removes the user “jane”.
 
+## Setting up SSL Keystore in Gateway
 
-## Enabling HTTPS in Gateway
+An SSL keystore is needed for your gateway to enable HTTPS in the gateway and to configure SMTP with password authentication. SMTP configuration is needed for the gateway to
```
Comment (amberarrow) [inline docs/dtgateway_security.md]: There is some discussion of this in the section entitled "Enabling HTTPS in Gateway" at http://docs.datatorrent.com/dtgateway_security/; please add a link to it.

File: docs/dtgateway_security.md [hunk 2]
```diff
@@ -821,34 +821,70 @@ This command returns the information about the user “john”.
...
           </property>
           <property>
-                    <name>dt.gateway.sslKeystorePassword</name>
-                     <value>{keystore-password}</value>
+            <!-- this is the keystore password -->
+              <name>dt.gateway.sslKeystorePassword</name>
+              <value>{keystore-password}</value>
           </property>
+        ...
+        </configuration>
```
Comment (amberarrow) [inline docs/dtgateway_security.md]: Seems duplicative of the link noted above

File: docs/dtgateway_security.md [hunk 3]
```diff
@@ -821,34 +821,70 @@ This command returns the information about the user “john”.
 
 This command removes the user “jane”.
 
+## Setting up SSL Keystore in Gateway
 
-## Enabling HTTPS in Gateway
+An SSL keystore is needed for your gateway to enable HTTPS in the gateway and to configure SMTP with password authentication. SMTP configuration is needed for the gateway to 
+to send email alerts as described in [System Alerts](dtgateway_systemalerts.md).
```
Comment (amberarrow) [inline docs/dtgateway_security.md]: omit "to"

File: docs/dtgateway_security.md [hunk 4]
```diff
@@ -821,34 +821,70 @@ This command returns the information about the user “john”.
...
 
-## Enabling HTTPS in Gateway
+An SSL keystore is needed for your gateway to enable HTTPS in the gateway and to configure SMTP with password authentication. SMTP configuration is needed for the gateway to 
+to send email alerts as described in [System Alerts](dtgateway_systemalerts.md). 
 
-HTTPS in the Gateway can be enabled by performing following two steps.
+Follow the steps below to set up the keystore.
 
-1.  Generate an SSL keystore if you don’t have one.  Instruction on how to generate an SSL keystore is here: [http://docs.oracle.com/cd/E19509-01/820-3503/ggfen/index.html](http://docs.oracle.com/cd/E19509-01/820-3503/ggfen/index.html)
+1. Generate an SSL keystore if you don’t have one.  Instruction on how to generate an SSL keystore is here: [http://docs.oracle.com/cd/E19509-01/820-3503/ggfen/index.html](http://docs.oracle.com/cd/E19509-01/820-3503/ggfen/index.html). Note down the full path of the keystore and the keystore password that you need to provide in the next step.
```
Comment (amberarrow) [inline docs/dtgateway_security.md]: Suggest switching the order: 
full path of the keystore and the keystore password => keystore password and the full path of the keystore

File: docs/dtgateway_security.md [hunk 5]
```diff
@@ -821,34 +821,70 @@ This command returns the information about the user “john”.
...
 
-2.  Add a property to dt-site.xml configuration file, typically located under `/opt/datatorrent/current/conf` ( or `~/datatorrent/current/conf` for local install).
+2. Add two properties to dt-site.xml configuration file, typically located under `/opt/datatorrent/current/conf` ( or `~/datatorrent/current/conf` for local install).
```
Comment (amberarrow) [inline docs/dtgateway_security.md]: dt-site.xml => `dt-site.xml`

File: docs/dtgateway_security.md [hunk 6]
```diff
@@ -821,34 +821,70 @@ This command returns the information about the user “john”.
...
+        ...
+        </configuration>
+
+3. Perform any additional steps required for "Enabling HTTPS" or "SMTP Password Encryption" use-case as described below.
+
+4. Restart the Gateway by running
+
+        sudo service dtgateway restart
+
+    ( when running Gateway in local mode use `dtgateway restart` command)
```
Comment (amberarrow) [inline docs/dtgateway_security.md]: ( when => (when

File: docs/dtgateway_security.md [hunk 7]
```diff
@@ -821,34 +821,70 @@ This command returns the information about the user “john”.
...
+
+### Enabling HTTPS in Gateway
+
+To enable HTTPS in the Gateway after setting up the keystore as described above, you have to add the following
+property to the dt-site.xml configuration file
```
Comment (amberarrow) [inline docs/dtgateway_security.md]: dt-site.xml => `dt-site.xml`

File: docs/dtgateway_security.md [hunk 8]
```diff
@@ -821,34 +821,70 @@ This command returns the information about the user “john”.
...
+
+        <configuration>
+        ...
           <property>
             <name>dt.attr.GATEWAY_USE_SSL</name>
                   <value>true</value>
```
Comment (amberarrow) [inline docs/dtgateway_security.md]: indentation

File: docs/dtgateway_systemalerts.md [hunk 9]
```diff
@@ -62,6 +62,80 @@ exist for every running application. The alert condition can refer to multiple s
...
+a note of the following steps before attempting to configure SMTP in the gateway.
+
+ * make sure there is an SMTP server in the network the gateway will have access to. Note down the
+SMTP server's hostname and port number (please see the note below regarding encryption-type).
+ * if the SMTP server supports or requires [TLS or SSL encryption](https://www.fastmail.com/help/technical/ssltlsstarttls.html),
+determine the encryption-type for the communication between the gateway and the SMTP server. The values for
+encryption-type are "tls", "ssl" or "none" (i.e. no encryption). In case "tls" or "ssl" is used,
+ensure appropriate certificates are installed to establish [trust](https://en.wikipedia.org/wiki/Public_key_certificate).
+ * determine the authentication-type for your SMTP server. The only supported values are "password" (i.e. authentication using
+username and password) and "none" (i.e. no authentication).
```
Comment (amberarrow) [inline docs/dtgateway_systemalerts.md]: and => or {to be consistent with the rest of the text}

File: docs/dtgateway_systemalerts.md [hunk 10]
```diff
@@ -62,6 +62,80 @@ exist for every running application. The alert condition can refer to multiple s
...
+ * if you choose the "password" authentication-type, you will need the SMTP-username and SMTP-password to be used
+by the gateway to authenticate with the SMTP server.
+ * determine the "from-address" and the "from-name" for the emails sent by the gateway which the email recipient
+will see as the sender of the emails. Note that for certain SMTP servers the "from-address" and the SMTP-username
+may need to match or the SMTP-username overrides the "from-address".
```
Comment (amberarrow) [inline docs/dtgateway_systemalerts.md]: or the SMTP-username overrides the "from-address". => or else the latter overrides the former.

File: docs/dtgateway_systemalerts.md [hunk 11]
```diff
@@ -62,6 +62,80 @@ exist for every running application. The alert condition can refer to multiple s
...
+SMTP server's hostname and port number (please see the note below regarding encryption-type).
+ * if the SMTP server supports or requires [TLS or SSL encryption](https://www.fastmail.com/help/technical/ssltlsstarttls.html),
+determine the encryption-type for the communication between the gateway and the SMTP server. The values for
+encryption-type are "tls", "ssl" or "none" (i.e. no encryption). In case "tls" or "ssl" is used,
+ensure appropriate certificates are installed to establish [trust](https://en.wikipedia.org/wiki/Public_key_certificate).
+ * determine the authentication-type for your SMTP server. The only supported values are "password" (i.e. authentication using
+username and password) and "none" (i.e. no authentication).
+ * if you choose the "password" authentication-type, you will need the SMTP-username and SMTP-password to be used
+by the gateway to authenticate with the SMTP server.
+ * determine the "from-address" and the "from-name" for the emails sent by the gateway which the email recipient
```
Comment (amberarrow) [inline docs/dtgateway_systemalerts.md]: These literal names do not match those in the PUT example below.

File: docs/dtgateway_systemalerts.md [hunk 12]
```diff
@@ -62,6 +62,80 @@ exist for every running application. The alert condition can refer to multiple s
...
+"fromAddr": "no-reply@mydomain.com",
+"fromName": "Do Not Reply",
+"encryptionType": "tls",
+"authType": "password",
+"username": "smtp-user@mydomain.com",
+"password": "secret_password",
+"testAddr": "testuser@yourdomain.com"
+}
+```
+
```
Comment (amberarrow) [inline docs/dtgateway_systemalerts.md]: Use a table for the descriptions below for better readability.

File: docs/dtgateway_systemalerts.md [hunk 13]
```diff
@@ -62,6 +62,80 @@ exist for every running application. The alert condition can refer to multiple s
...
+will see as the sender of the emails. Note that for certain SMTP servers the "from-address" and the SMTP-username
+may need to match or the SMTP-username overrides the "from-address".
+ * determine a "test-address" where you can receive emails to verify validity of the SMTP configuration.
+ * in case you choose the "password" authentication-type, you will need to set up an SSL keystore as described
+[here](dtgateway_security.md#setting-up-ssl-keystore-in-gateway). The gateway uses the keystore to encrypt and
+decrypt your SMTP-password.
+
+You can use the following APIs to set or retrieve the SMTP configuration.
+
+### PUT /ws/v2/config/smtp
```
Comment (amberarrow) [inline docs/dtgateway_systemalerts.md]: Clarify that the following JSON sample is the payload of the PUT command

File: docs/dtgateway_systemalerts.md [hunk 14]
```diff
@@ -62,6 +62,80 @@ exist for every running application. The alert condition can refer to multiple s
...
+username = the "SMTP-username" value described above
+password = the "SMTP-password" value described above
+testAddr = the "test-address" value described above. The gateway sends a test email to this address when you use this API to set the SMTP configuration.
+
+A note about password: if you omit the "password" value in your JSON object in the API, the gateway will use the existing saved password
+so the client does not need to include the password in subsequent API calls. Note that the GET API (described below) never returns the
+password of the SMTP configuration, hence the Datatorrent console is able to use this feature without either displaying or requiring the 
+user to re-enter the previous password.
+
+### GET /ws/v2/config/smtp
```
Comment (amberarrow) [inline docs/dtgateway_systemalerts.md]: Clarify that the following JSON is the value returned from the GET command

**Comment by sanjaypujare** [comment]
@amberarrow incorporated comments

File: docs/dtgateway_systemalerts.md [hunk 15]
```diff
@@ -62,6 +62,85 @@ exist for every running application. The alert condition can refer to multiple s
...
+by the gateway to authenticate with the SMTP server.
+ * determine the "fromAddr" and the "fromName" for the emails sent by the gateway which the email recipient
+will see as the sender of the emails. Note that for certain SMTP servers the "fromAddr" and the SMTP-username
+may need to match or else the latter overrides the former.
+ * determine a "test-address" where you can receive emails to verify validity of the SMTP configuration.
+ * in case you choose the "password" authentication-type, you will need to set up an SSL keystore as described
+[here](dtgateway_security.md#setting-up-ssl-keystore-in-gateway). The gateway uses the keystore to encrypt and
+decrypt your SMTP-password.
+
+You can use the following APIs to set or retrieve the SMTP configuration. Note that the JSON sample folllowing the
```
Comment (amberarrow) [inline docs/dtgateway_systemalerts.md]: folllowing => following

File: docs/dtgateway_systemalerts.md [hunk 16]
```diff
@@ -62,6 +62,85 @@ exist for every running application. The alert condition can refer to multiple s
...
+|fromName|the "from-name" described above i.e. the user friendly string for the "from-address"|
+|encryptionType|the "encryption-type" value described above i.e. "tls", "ssl" or "none"|
+|authType|the "authentication-type" value described above i.e. "password" or "none"|
+|username|the "SMTP-username" value described above|
+|password|the "SMTP-password" value described above|
+|testAddr|the "test-address" value described above. The gateway sends a test email to this address when you use this API to set the SMTP configuration.|
+
+A note about password: if you omit the "password" value in your JSON object in the API, the gateway will use the existing saved password
+so the client does not need to include the password in subsequent API calls. Note that the GET API (described below) never returns the
+password of the SMTP configuration, hence the Datatorrent console is able to use this feature without either displaying or requiring the
```
Comment (amberarrow) [inline docs/dtgateway_systemalerts.md]: Datatorrent => DataTorrent

File: docs/dtgateway_security.md [hunk 17]
```diff
@@ -821,34 +821,71 @@ This command returns the information about the user “john”.
...
-		
-    ( when running Gateway in local mode use `dtgateway restart` command)
+Add another key to the keystore created above using the instructions mentioned in [http://docs.oracle.com/cd/E19509-01/820-3503/ggfen/index.html](http://docs.oracle.com/cd/E19509-01/820-3503/ggfen/index.html). For example:
+
+keytool -genkey -alias smtpenc-alias -keyalg RSA -keypass <key password> -storepass <store password> -keystore gwkeystore.jks
+
+Note down the alias you used for the key (smtpenc-alias in the above command). Remember to use the same key password and 
+store password as the ones you used in the previous command where you created the keystore.
+
+Add the following property to dt-site.xml to indicate the alias to be used for SMTP Password Encryption:
```
Comment (amberarrow) [inline docs/dtgateway_security.md]: dt-site.xml => `dt-site.xml`

File: docs/dtgateway_security.md [hunk 18]
```diff
@@ -821,34 +821,71 @@ This command returns the information about the user “john”.
...
 
-3.  Restart the Gateway by running
+### Setting up a Key for SMTP Password Encryption
 
-        sudo service dtgateway restart
-		
-    ( when running Gateway in local mode use `dtgateway restart` command)
+Add another key to the keystore created above using the instructions mentioned in [http://docs.oracle.com/cd/E19509-01/820-3503/ggfen/index.html](http://docs.oracle.com/cd/E19509-01/820-3503/ggfen/index.html). For example:
+
+keytool -genkey -alias smtpenc-alias -keyalg RSA -keypass <key password> -storepass <store password> -keystore gwkeystore.jks
```
Comment (amberarrow) [inline docs/dtgateway_security.md]: Use a triple backquoted block or indent 4 spaces so it looks better.

**Comment by amberarrow** [comment]
Found a few more minor items (sorry for not catching them earlier).

File: docs/dtgateway_systemalerts.md [hunk 19]
```diff
@@ -62,6 +62,85 @@ exist for every running application. The alert condition can refer to multiple s
...
+|encryptionType|the "encryption-type" value described above i.e. "tls", "ssl" or "none"|
+|authType|the "authentication-type" value described above i.e. "password" or "none"|
+|username|the "SMTP-username" value described above|
+|password|the "SMTP-password" value described above|
+|testAddr|the "test-address" value described above. The gateway sends a test email to this address when you use this API to set the SMTP configuration.|
+
+A note about password: if you omit the "password" value in your JSON object in the API, the gateway will use the existing saved password
+so the client does not need to include the password in subsequent API calls. Note that the GET API (described below) never returns the
+password of the SMTP configuration, hence the Datatorrent console is able to use this feature without either displaying or requiring the 
+user to re-enter the previous password. Note that the JSON sample folllowing the GET request is the value returned from the GET request.
```
Comment (amberarrow) [inline docs/dtgateway_systemalerts.md]: folllowing => following

File: docs/dtgateway_systemalerts.md (hunk 12 above)
Comment (sanjaypujare) [inline docs/dtgateway_systemalerts.md]: done.

**Comment by sanjaypujare** [comment]
incorporated comments.

**Pull request merged**